brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -6,81 +6,145 @@ import com.twitter.algebird._
+          parent <- tree.leafAt(leafIndex).toList;
+          (feature, stats) <- leafStats.toList;
+          split <- splitter.split(parent.target, stats);
+          (newSplit, score) <- evaluator.evaluate(split).toList
+        ) yield (newSplit.createSplitNode(feature), score)
+
+        if(candidates.isEmpty)
+          None
+        else
+          Some(candidates.maxBy{_._2}._1)
+      }
     }
+
+    val newTrees = 0.until(trees.size).toList.map{i => newTreeMap(i)}
+    Trainer(trainingData, sampler, newTrees)
   }
 
-  def updateTargets: Trainer[K, V, T] =
-    updateLeaves {
-      case (treeIndex, (index, _, annotation), instances) =>
-        val target = implicitly[Monoid[T]].sum(instances.map { _.target })
-        LeafNode(index, target, annotation)
+  def updateTargets: Trainer[K, V, T] = {
+    var targets = treeMap.mapValues{tree => Map[Int, T]()}
+    trainingData.foreach{instance =>
+      for (
+        (treeIndex, tree) <- treeMap.toList;
+        i <- 1.to(sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)).toList;
+        leafIndex <- tree.leafIndexFor(instance.features).toList
+      ) {
+        val treeTargets = targets(treeIndex)
+        val old = treeTargets.getOrElse(leafIndex, Monoid.zero[T])
+        val combined = Monoid.plus(instance.target, old)
+        targets += treeIndex -> (treeTargets + (leafIndex -> combined))
+      }
     }
 
-  def expand(times: Int)(implicit splitter: Splitter[V, T], evaluator: Evaluator[V, T], stopper: Stopper[T]): Trainer[K, V, T] =
-    updateLeaves {
-      case (treeIndex, (index, target, annotation), instances) =>
-        Tree.expand(times, treeIndex, LeafNode[K, V, T, Unit](index, target, annotation), splitter, evaluator, stopper, sampler, instances)
+    val newTrees = trees.zipWithIndex.map{case (tree, index) =>
+      tree.updateByLeafIndex{leafIndex =>
+        val target = targets(index).getOrElse(leafIndex, Monoid.zero[T])
+        Some(LeafNode(leafIndex, target, ()))
+      }
     }
 
-  def prune[P, E](error: Error[T, P, E])(implicit voter: Voter[T, P], ord: Ordering[E]): Trainer[K, V, T] =
-    updateTrees {
-      case (tree, treeIndex, byLeaf) =>
-        val byLeafIndex = byLeaf.map {
-          case ((index, _, _), instances) =>
-            index -> implicitly[Monoid[T]].sum(instances.map { _.target })
-        }
-        tree.prune(byLeafIndex, voter, error)
-    }
+    copy(trees = newTrees)
+  }
 
   def validate[P, E](error: Error[T, P, E])(implicit voter: Voter[T, P]): Option[E] = {
-    val errors = trainingData.flatMap { instance =>
-      val useTrees = trees.zipWithIndex.filter {
-        case (tree, i) =>
-          sampler.includeInValidationSet(instance.id, instance.timestamp, i)
-      }.map { _._1 }
-      if(useTrees.isEmpty)
-        None
-      else {
-        val prediction = voter.predict(useTrees, instance.features)
-        Some(error.create(instance.target, prediction))
+    var output: Option[E] = None
```

Comment:
Do we expect that we won't have a monoid for error? It seems like using `Monoid[E]` here would be preferable unless that was a non-starter.